### PR TITLE
feat: integrate application session management into Casdoor's session management (#774)

### PR DESCRIPTION
**Pull request by xiao-niu-ren**

### CODE REVIEW

**Comment by hsluoyz** [comment]
@leo220yuyaodog plz review

File: object/session.go [hunk 1]
```diff
@@ -20,25 +20,37 @@ import (
...
 	Owner       string `xorm:"varchar(100) notnull pk" json:"owner"`
 	Name        string `xorm:"varchar(100) notnull pk" json:"name"`
+	Application string `xorm:"varchar(100) notnull pk default ''" json:"application"`
 	CreatedTime string `xorm:"varchar(100)" json:"createdTime"`
 
 	SessionId []string `json:"sessionId"`
 }
 
-func SetSession(id string, sessionId string) {
+func SetSession(id string, sessionId string, application *Application) {
```
Comment (hsluoyz) [inline object/session.go:37]: UpdateSession

File: object/session.go [hunk 2]
```diff
@@ -48,26 +60,26 @@ func SetSession(id string, sessionId string) {
 	}
 }
 
-func DeleteSession(id string) bool {
+func DeleteSession(id string, applicationName string) bool {
```
Comment (hsluoyz) [inline object/session.go:63]: applicationName is added to PK, it's already a part of id now

File: object/session.go [hunk 3]
```diff
@@ -48,26 +60,26 @@ func SetSession(id string, sessionId string) {
...
 	}
 
 	DeleteBeegoSession(session.SessionId)
 
-	affected, err := adapter.Engine.ID(core.PK{owner, name}).Delete(session)
+	affected, err := adapter.Engine.ID(core.PK{owner, name, applicationName}).Delete(session)
 	return affected != 0
 }
 
 func DeleteSessionId(id string, sessionId string) bool {
```
Comment (hsluoyz) [inline object/session.go:78]: DeleteSession

File: object/session.go [hunk 4]
```diff
@@ -128,3 +140,44 @@ func GetSessionCount(owner, field, value string) int {
 
 	return int(count)
 }
+
+func AddUserSession(owner string, applicationName string, userName string, sessionId []string, sessionCreateTime string) bool {
```
Comment (hsluoyz) [inline object/session.go:144]: AddSession

File: object/session.go [hunk 5]
```diff
@@ -128,3 +140,44 @@ func GetSessionCount(owner, field, value string) int {
...
+		session.SessionId = append(session.SessionId, v)
+	}
+	affected, err := adapter.Engine.Insert(session)
+	if err != nil {
+		panic(err)
+	}
+	return affected != 0
+}
+
+func DeleteUserSession(owner string, applicationName string, userName string) bool {
```
Comment (hsluoyz) [inline object/session.go:158]: DeleteSession

File: object/session.go [hunk 6]
```diff
@@ -128,3 +140,44 @@ func GetSessionCount(owner, field, value string) int {
...
+	_, err := adapter.Engine.ID(core.PK{owner, userName, applicationName}).Get(session)
+	if err != nil {
+		return false
+	}
+
+	affected, err := adapter.Engine.ID(core.PK{owner, userName, applicationName}).Delete(session)
+	return affected != 0
+}
+
+func IsUserSessionDuplicated(owner string, applicationName string, userName string, sessionId string) bool {
```
Comment (hsluoyz) [inline object/session.go:169]: IsSessionDuplicated

File: routers/router.go [hunk 7]
```diff
@@ -165,6 +165,10 @@ func initAPI() {
 	beego.Router("/api/get-sessions", &controllers.ApiController{}, "GET:GetSessions")
 	beego.Router("/api/delete-session", &controllers.ApiController{}, "POST:DeleteSession")
 
+	beego.Router("/api/add-user-session", &controllers.ApiController{}, "POST:AddUserSession")
+	beego.Router("/api/delete-user-session", &controllers.ApiController{}, "POST:DeleteUserSession")
+	beego.Router("/api/is-user-session-duplicated", &controllers.ApiController{}, "GET:IsUserSessionDuplicated")
+
```
Comment (hsluoyz) [inline routers/router.go:171]: Finally will be 6 APIs:

```go
beego.Router("/api/get-sessions", &controllers.ApiController{}, "GET:GetSessions")
beego.Router("/api/get-session", &controllers.ApiController{}, "GET:GetSession")
beego.Router("/api/update-session", &controllers.ApiController{}, "POST:UpdateSession")
beego.Router("/api/add-session", &controllers.ApiController{}, "POST:AddSession")
beego.Router("/api/delete-session", &controllers.ApiController{}, "POST:DeleteSession")
beego.Router("/api/is-session-duplicated", &controllers.ApiController{}, "GET:IsSessionDuplicated")
```

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/session.go [hunk 8]
```diff
@@ -20,25 +20,37 @@ import (
...
 
+var (
+	casdoorApplication  = "app-built-in"
+	casdoorOrganization = "built-in"
+)
+
 type Session struct {
 	Owner       string `xorm:"varchar(100) notnull pk" json:"owner"`
 	Name        string `xorm:"varchar(100) notnull pk" json:"name"`
+	Application string `xorm:"varchar(100) notnull pk default ''" json:"application"`
```
Comment (leo220yuyaodog) [inline object/session.go:31]: Why application is primery key?

File: object/session.go (hunk 4 above)
Comment (leo220yuyaodog) [inline object/session.go:144]: Don't need parameter `sessionCreateTime`, it can get in this function.

File: controllers/session.go [hunk 9]
```diff
@@ -66,3 +66,65 @@ func (c *ApiController) GetSessions() {
 		c.ResponseOk(sessions, paginator.Nums())
 	}
 }
+
+// AddUserSession
```
Comment (leo220yuyaodog) [inline controllers/session.go:70]: What is the use of api addSession? We can not request the interface to add a session, only after the backend beego created, then casdoor to add a session

File: object/session.go [hunk 10]
```diff
@@ -93,38 +178,19 @@ func DeleteBeegoSession(sessionIds []string) {
 	}
 }
 
-func GetSessions(owner string) []*Session {
-	sessions := []*Session{}
-	var err error
-	if owner != "" {
-		err = adapter.Engine.Desc("created_time").Where("owner = ?", owner).Find(&sessions)
+func IsSessionDuplicated(id string, sessionId string) bool {
+	owner, name, application := util.GetOwnerAndNameAndApplicationFromSessionPkId(id)
```
Comment (hsluoyz) [inline object/session.go:182]: GetOwnerAndNameAndOtherFromId

File: util/string.go [hunk 11]
```diff
@@ -100,6 +100,15 @@ func GetOwnerAndNameFromIdNoCheck(id string) (string, string) {
 	return tokens[0], tokens[1]
 }
 
+func GetOwnerAndNameAndApplicationFromSessionPkId(id string) (string, string, string) {
```
Comment (hsluoyz) [inline util/string.go:103]: GetOwnerAndNameAndOtherFromId

File: util/string.go [hunk 12]
```diff
@@ -127,6 +136,10 @@ func GetId(owner, name string) string {
 	return fmt.Sprintf("%s/%s", owner, name)
 }
 
+func GetSessionPkId(owner, name, application string) string {
```
Comment (hsluoyz) [inline util/string.go:139]: GetSessionId

File: object/session.go [hunk 13]
```diff
@@ -20,66 +20,151 @@ import (
...
 
+var (
+	casdoorApplication  = "app-built-in"
+	casdoorOrganization = "built-in"
+)
+
 type Session struct {
 	Owner       string `xorm:"varchar(100) notnull pk" json:"owner"`
 	Name        string `xorm:"varchar(100) notnull pk" json:"name"`
+	Application string `xorm:"varchar(100) notnull pk default ''" json:"application"`
```
Comment (hsluoyz) [inline object/session.go:31]: Remove `default ''`

File: object/session.go [hunk 14]
```diff
@@ -20,66 +20,151 @@ import (
 	"xorm.io/core"
 )
 
+var (
+	casdoorApplication  = "app-built-in"
+	casdoorOrganization = "built-in"
+)
+
 type Session struct {
 	Owner       string `xorm:"varchar(100) notnull pk" json:"owner"`
```
Comment (hsluoyz) [inline object/session.go:29]: You have changed the PK of session table, please handle DB migration well: https://casdoor.org/docs/deployment/db-migration

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/session.go [hunk 15]
```diff
@@ -66,3 +47,93 @@ func (c *ApiController) GetSessions() {
...
 }
+
+// GetSingleSession
+// @Title GetSingleSession
+// @Tag Session API
+// @Description Get session for one user in one application.
+// @Param   ID     query    string  true        "The ID(organization/application/user) of session"
+// @Success 200 {array} string The Response object
+// @router /get-session [get]
+func (c *ApiController) GetSingleSession() {
```
Comment (hsluoyz) [inline controllers/session.go:58]: You don't need to rename. The original function in controllers can be renamed to `GetBeegoSession()`, yours can be `GetSession()`

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/session.go (hunk 9 above)
Comment (hsluoyz) [reply controllers/session.go]: It's for application to create a session at Casdoor

File: object/session.go [hunk 16]
```diff
@@ -20,66 +20,151 @@ import (
 	"xorm.io/core"
 )
 
+var (
```
Comment (hsluoyz) [inline object/session.go:23]: Make it public

File: controllers/account.go [hunk 17]
```diff
@@ -254,7 +254,7 @@ func (c *ApiController) Logout() {
 
 	if accessToken == "" && redirectUri == "" {
 		c.ClearUserSession()
-		object.DeleteSessionId(user, c.Ctx.Input.CruSession.SessionID())
+		object.DeleteSessionId(util.GetSessionId("built-in", "app-built-in", user), c.Ctx.Input.CruSession.SessionID())
```
Comment (hsluoyz) [inline controllers/account.go:257]: Use object pkg's 

```go
var (
	CasdoorApplication  = "app-built-in"
	CasdoorOrganization = "built-in"
)
```

File: controllers/account.go [hunk 18]
```diff
@@ -291,7 +291,7 @@ func (c *ApiController) Logout() {
 			}
 
 			c.ClearUserSession()
-			object.DeleteSessionId(user, c.Ctx.Input.CruSession.SessionID())
+			object.DeleteSessionId(util.GetSessionId("built-in", "app-built-in", user), c.Ctx.Input.CruSession.SessionID())
```
Comment (hsluoyz) [inline controllers/account.go:294]: Use object pkg's 

```go
var (
	CasdoorApplication  = "app-built-in"
	CasdoorOrganization = "built-in"
)
```

File: controllers/auth.go [hunk 19]
```diff
@@ -139,8 +139,13 @@ func (c *ApiController) HandleLoggedIn(application *object.Application, user *ob
 		})
 	}
 
-	if resp.Status == "ok" {
-		object.SetSession(user.GetId(), c.Ctx.Input.CruSession.SessionID())
+	if resp.Status == "ok" && user.Owner == "built-in" && application.Name == "app-built-in" {
```
Comment (hsluoyz) [inline controllers/auth.go:142]: Use object pkg's 

```go
var (
	CasdoorApplication  = "app-built-in"
	CasdoorOrganization = "built-in"
)
```

File: object/adapter.go [hunk 20]
```diff
@@ -223,12 +224,104 @@ func (a *Adapter) createTable() {
 		panic(err)
 	}
 
-	err = a.Engine.Sync2(new(Session))
+	err = a.syncSession()
```
Comment (hsluoyz) [inline object/adapter.go:227]: Don't change this

File: object/adapter.go [hunk 21]
```diff
@@ -223,12 +224,104 @@ func (a *Adapter) createTable() {
...
 	if err != nil {
 		panic(err)
 	}
 }
 
+func (a *Adapter) syncSession() error {
```
Comment (hsluoyz) [inline object/adapter.go:233]: Define migrate pkg, and define a `Migrator` struct.

- doMigration

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/account.go (hunk 18 above)
Comment (leo220yuyaodog) [inline controllers/account.go:294]: The api `logout` is also used for third-party application logout according openid protocol. Here should use `application.name` `application.organization`. 

If it's not the main content of the pr. I'll fixed it after the pr merged.

**Pull request closed without merging**